sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -121,98 +121,233 @@ Use ``node make help`` to list all available commands.
-fairly sensitive to system configuration (especially in regard to differences in
-font rendering). Visual tests can be performed locally, but given that baseline
-images for all three supported platforms have to be updated, the preferred approach
-is to generate images and compare them in CI.
-
-The full procedure for visual testing is as follows:
-
-1. Make changes to the repository and write new tests or update existing.
-2. Use ``node make tests`` to incrementally test your changes on your system.
-3. Push your changes to GitHub and wait for CI to finish.
-4. If you added new tests, CI will expectedly fail with "missing baseline
-   images" error message.
-5. If tests passed then you are done.
-6. If tests failed, go to BokehJS's GitHub_Actions_ page. Find the most recent
-   test run for your PR and download the associated ``bokehjs-report`` artifact.
-7. Unzip the artifact archive at the root of the repository.
-8. Assuming devtools server is running in the background, go to ``/integration/report?platform=name``
-   where ``name`` is either ``linux``, ``macos`` or ``windows`` and review the test output
-   for each platform. If there are no unintentional differences, then commit all
-   new or modified ``*.blf`` and ``*.png`` files under ``test/baselines/{linux,macos,windows}``.
-9. Push your changes to GitHub again and verify that tests pass this time.
-
-.. note::
-
-    Make sure to monitor the state of the ``test/baselines`` directory, so that you
-    don't commit unnecessary files. If you do so, subsequent tests will fail. Reset
-    this directory after every failed test run (``git checkout`` and/or ``git clean``).
+BokehJS uses a series of visual baseline comparison tests. These tests help make sure
+that Bokeh's visual output is consistent with the output expected by design. Any
+BokehJS-related pull requests that result in changes to the visual output generated by
+BokehJS should include visual baseline comparison tests.
+
+In the background, BokehJS' testing framework runs a headless browser and takes
+screenshots of the browser's output. The testing framework then compares the visual
+output to each test's dedicated baseline files.
+
+Each test in ``test:integration`` consists of two types of baseline comparisons:
+
+Textual baseline comparison
+  For each test, the testing framework compares the pixel location of certain elements
+  in the visual output to pixel locations in a baseline file. The pixel locations are a
+  set of bounding boxes. They are stored as plain text in each test's ``.blf`` files.
+
+Visual baseline comparison
+  For each test, the testing suite does a pixel-by-pixel comparison of a screenshot
+  and a baseline image. These baseline images are stored as ``.png`` files. In contrast
+  to textual baseline comparisons, visual baseline comparisons are platform-dependent.
+  Even minor differences in font rendering, for example, will make the pixel-by-pixel
+  comparison fail.
```

Comment:
If we are going into such details, then `it.allowing()` should be explained as well.